Sprint #22 looks normal to me but obviously not perfect )

nim0y opened the pull request.

nim0y committed:
Sprint #22 looks normal to me but obviously not perfect )

ArturNurtdinov commented on app/src/main/AndroidManifest.xml:
```diff
@@ -4,6 +4,10 @@
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission
+        android:name="android.permission.WRITE_EXTERNAL_STORAGE"
+        tools:ignore="ScopedStorage" />
```
Это лишнее. Чтобы пользоваться системным пикером тебе не нужен этот пермишен, к тому же он не имеет эффекта на версиях андроид выше 12

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/db/TrackToPlaylistEntity.kt:
```diff
@@ -0,0 +1,21 @@
+@Entity(tableName = TRACKS_IN_PLAYLIST)
+data class TrackToPlaylistEntity(
+    @PrimaryKey(autoGenerate = true)
+    val trackId: Int?,
```
id не стоит делать опциональным, т.к. это primary key в таблице

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/db/converters/TracksIdsConverter.kt:
```diff
@@ -0,0 +1,22 @@
+class TracksIdsConverter {
+    @TypeConverter
+    fun fromList(list: ArrayList<Long>): String {
+        return list.joinToString(",")
```
Сепараторы через запятую являются внутренней логикой этого маппера, стоит вынести это в константу

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/db/converters/TracksToPlaylistConverter.kt:
```diff
@@ -0,0 +1,39 @@
+            track.primaryGenreName,
+            track.country,
+            track.previewUrl,
+            Date().time
```
Время в мс лучше передавать как параметр функции сюда, т.к. маппер не должен отвечать за проставление этого поля, за это должен отвечать репозиторий

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/impl/db/PlaylistsRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    override suspend fun addTrackToPlaylist(playList: Playlist, track: Track) {
+        playList.tracksIds.add(track.trackId?.toLong() ?: 0)
```
id не должен быть нуллабельным в общем случае, но даже если так, то нельзя добавлять с id = 0, лучше вернуть ошибку

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/impl/db/PlaylistsRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+
+    override suspend fun saveCoverToPrivateStorage(previewUri: Uri, context: Context): Uri? {
+        val path =
+            File(context.getExternalFilesDir(Environment.DIRECTORY_PICTURES), "myPics")
```
Название директории стоит вынести в константы

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/db/PlaylistsRepository.kt:
```diff
@@ -0,0 +1,24 @@
+
+interface PlaylistsRepository {
+
+    suspend fun getPlaylists(): Flow<List<Playlist>>
```
Не имеет смысла делать функцию, возвращающую Flow, suspend. Либо одно, либо другое

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,114 @@
+import org.koin.androidx.viewmodel.ext.android.viewModel
+
+class NewPlaylistFragment : Fragment() {
+    private var _binding: FragmentCreatePlaylistBinding? = null
```
При использовании `viewBinding` во фрагментах обязательно нужно обнулять `_binding` в `onDestroyView`, иначе будет утечка памяти. Почитать о том, как использовать эту фичу во фрагментах можно тут - https://developer.android.com/topic/libraries/view-binding#fragments. Почитать подробнее про утечку можно здесь - https://stackoverflow.com/questions/65295104/android-view-binding-clear-binding-in-fragment-lifecycle

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,114 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        (activity as MainActivity).hideNavBar()
```
```suggestion
        (activity as? MainActivity)?.hideNavBar()
```
Safe cast предпочтительнее обычного каста почти во всех случаях

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,114 @@
+                findNavController().navigateUp()
+            }
+        }
+        requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner, callback)
```
Этот колбэк стоит удалять в функции `onDestroyView`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,114 @@
+                tracksAmount = 0,
+                imageUri = coverUri.toString()
+            )
+        )
```
Созданием плейлиста должен заниматься репозиторий, это не ответственность фрагмента. В репозиторий нужно прокинуть данные для создания плейлиста

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/NewPlaylistViewModel.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun getCover(): String {
+        return "cover_${UUID.randomUUID()}.jpg"
+    }
```
Это логика data слоя, не presentation

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/PlayListFragment.kt:
```diff
@@ -24,13 +32,81 @@ class PlayListFragment : Fragment() {
+//        val bundle = Bundle().apply {
+//            putParcelable("Put_Playlist",playlist)
+//        }
+//        findNavController().navigate() -future playlist +bundle
```
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/PlaylistsViewHolder.kt:
```diff
@@ -0,0 +1,44 @@
+    fun bind(model: Playlist, clickListener: ClickListener) {
+        binding.namePlaylist.text = model.name
+        binding.descriptionPlaylist.text = setTracksAmount(itemView.context, model.tracksAmount)
+        val previewUri = model.imageUri.let { Uri.parse(it) }
```
Парсинг Uri следует делать во вьюмодели. Если делать это на ui слое, то возможны проблемы с перформансом, т.к. парсинг это довольно тяжёлая операция

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/PlaylistsViewHolder.kt:
```diff
@@ -0,0 +1,44 @@
+            else -> context.getString(R.string.track_multiplie)
+        }
+        return "$count $countTrack"
+    }
```
В андроид уже есть plurals - https://developer.android.com/guide/topics/resources/string-resource#Plurals, стоит пользоваться ими и не изобретать велосипед, т.к. при поддержке других локалей эта функция сильно разрастётся

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/playlists/PlaylistsViewHolder.kt:
```diff
@@ -0,0 +1,44 @@
+        return "$count $countTrack"
+    }
+
+    fun interface ClickListener {
+        fun onClick(playlist: Playlist)
+    }
```
Можно просто лямбду использовать, без интерфейса

ArturNurtdinov reviewed (commented):
1. При выборе фото при создании плейлиста область для фото не имеет закругленных углов. Отступ между полями для ввода названия и описания не соответствует дизайну:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/fc688ff7-f2df-4af9-a3f3-03e53ff5c665)
ОР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/e068ffba-4a4a-4a67-ab0a-2649977bc671)
2. Бэкграунд плейлиста без обложки не соответствует дизайну в тёмной теме:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/daf58f5f-93fd-4daa-a6f5-2aef5ed89b09)
ОР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/def2f8a8-6866-4567-92b0-6646bb93e65c)
3. Бэкграунд на экране создания плейлиста не соответствует дизайну в тёмной теме:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/a39a3537-307c-430d-8606-0e6d4d604648)
ОР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/dddb14b5-0da6-41a7-a1bd-558e98ebce47)
Цвет должен быть #1A1B22
4. Отступ кнопки "Новый плейлист" от начала списка плейлистов не соответствует дизайну. Должен быть 16dp:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/65530220-b90b-4234-9eaa-ce23c0478f36)
ОР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/cfd5bfc8-1bfa-4df0-91bf-66755fb9679d)

nim0y committed:
Sprint #22 second try

nim0y committed:
Sprint #22 second try

nim0y committed:
Sprint #22 ///

ArturNurtdinov reviewed (commented):
1. Диалог не соответствует дизайну в темной теме, судя по дизайну стиль диалога в тёмной теме должен оставаться таким же, как в светлой:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/b04ab4a0-fdab-4e4e-b187-ed867bb28d1b)
ОР:
*Пример диалога из следующего спринта в тёмной теме*
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/cbea13ff-2ad3-4eb6-8215-adf35035c626)

ArturNurtdinov commented:
Бэкграунд плейлиста без обложки не соответствует дизайну в тёмной теме:
ФР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/f855fd43-84fc-4bd0-9b6c-ff3911f484cd)
ОР:
![image](https://github.com/nim0y/PlaylistMaker/assets/49075521/bf15d7c2-fc68-496b-8624-a5bae2305747)

nim0y committed:
Sprint #22 3rd try AlertDialog
В ссылке на Figma нет скрина с этим диалогом.
И в задании его то же нет.
Создал кастомную тему для AlertDialog ориентируясь по звёздам. (и твоему скрину :)
